src/Altinn.Platform/Altinn.Platform.Storage/Storage/Controllers/InstancesController.cs

```diff
@@ -323,16 +345,152 @@ public async Task<ActionResult> Post(string appId, int? instanceOwnerId, [FromBo
+            {
+                MediaTypeHeaderValue mediaType = MediaTypeHeaderValue.Parse(Request.ContentType);
+                string boundary = MultipartRequestHelper.GetBoundary(mediaType, _defaultFormOptions.MultipartBoundaryLengthLimit);
+
+                MultipartReader reader = new MultipartReader(boundary, Request.Body);
+
+                // Execute ReadNextSectionAsync() twice, to expose the second part of the Multipart content,
+                // first part has already been handled in ReadInstanceTemplateFromBody(HttpRequest req).
+                MultipartSection section = reader.ReadNextSectionAsync().Result;
+                section = reader.ReadNextSectionAsync().Result;
+
+                bool hasContentDisposition = ContentDispositionHeaderValue.TryParse(section.ContentDisposition, out ContentDispositionHeaderValue contentDisposition);
+
+                if (hasContentDisposition)
+                {
+                    contentFileName = contentDisposition.FileName.ToString();
+                    fileSize = contentDisposition.Size ?? 0;
+                }
+
+                while (section != null)
+                {
+                    string elementType = section.GetContentDispositionHeader().Name.ToString();
+
+                    if (!appInfo.ElementTypes.Exists(e => e.Id == elementType))
+                    {
+                        return BadRequest("Requested element type is not declared in application metadata");
+                    }
+                    
+                    if (section.ContentType.StartsWith("application/xml"))
+                    {
+                        theStream = section.Body;
+                        contentType = section.ContentType;
+                        
+                        // Create a new DataElement to store in blob within the Instance instance.
+                        DataElement newDataElement = _dataController.CreateDataElementHelper(elementType, createdInstance, Guid.Parse(createdInstance.Id), creationTime, contentType, contentFileName, fileSize);
+                        
+                        try
+                        {
+                            // store file as blob
+                            await _dataRepository.CreateDataInStorage(theStream, newDataElement.StorageUrl);
+
+                            createdInstance.Data.Add(newDataElement);
+                        }
+                        catch (Exception e)
+                        {
+                            return StatusCode(500, $"Unable to create instance data in storage: {e}");
+                        }
+                    }
+
+                    // Drains any remaining section body that has not been consumed and
+                    // reads the headers for the next section.
+                    section = reader.ReadNextSectionAsync().Result;
+                }
+            }
+            else
+            {
```

Comment:
This must be wrong. Isn't the multipart read logic going to be contained in the while loop.